<a href="https://colab.research.google.com/github/JIN-HAN2025/AICE-python-practice/blob/main/04_20251228_%ED%8C%90%EB%8B%A4%EC%8A%A4%EB%8D%B0%EC%9D%B4%ED%84%B0%EA%B0%80%EA%B3%B5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [5]:
#데이터 불러오기, 데이터 저장하기, 데이터 가공(변환)
import pandas as pd
df = pd.read_csv('/content/sample_data/iris.csv')
# print(df)
df

sepal.length  sepal.width  petal.length  petal.width    variety
0             5.1          3.5           1.4          0.2     Setosa
1             4.9          3.0           1.4          0.2     Setosa
2             4.7          3.2           1.3          0.2     Setosa
3             4.6          3.1           1.5          0.2     Setosa
4             5.0          3.6           1.4          0.2     Setosa
..            ...          ...           ...          ...        ...
145           6.7          3.0           5.2          2.3  Virginica
146           6.3          2.5           5.0          1.9  Virginica
147           6.5          3.0           5.2          2.0  Virginica
148           6.2          3.4           5.4          2.3  Virginica
149           5.9          3.0           5.1          1.8  Virginica

[150 rows x 5 columns]

In [10]:


#df = pd.read_csv('/content/sample_data/iris_noheader.csv',header=None)

#열이름 지정
col_names = ['ColA','ColB','ColC','ColD','ColE']
df = pd.read_csv('/content/sample_data/iris_noheader.csv',names = col_names)
df


ColA  ColB  ColC  ColD       ColE
0     5.1   3.5   1.4   0.2     Setosa
1     4.9   3.0   1.4   0.2     Setosa
2     4.7   3.2   1.3   0.2     Setosa
3     4.6   3.1   1.5   0.2     Setosa
4     5.0   3.6   1.4   0.2     Setosa
..    ...   ...   ...   ...        ...
145   6.7   3.0   5.2   2.3  Virginica
146   6.3   2.5   5.0   1.9  Virginica
147   6.5   3.0   5.2   2.0  Virginica
148   6.2   3.4   5.4   2.3  Virginica
149   5.9   3.0   5.1   1.8  Virginica

[150 rows x 5 columns]

In [13]:
# 엔코딩 지정
df_fire = pd.read_csv('/content/sample_data/2019년화재.csv',encoding='cp949') #cp949 유니코드 , 지정안하면 utf-8이 기본으로 설정
df_fire

연번  사망  부상  인명피해(명)소계  재산피해소계           화재발생년월일       시도   시군구  \
0          1   0   1          1       0   2019-01-01 0:03      강원도   삼척시   
1          2   0   0          0   14900   2019-01-01 0:06      경기도   평택시   
2          3   0   0          0    3642   2019-01-01 0:40    서울특별시   성북구   
3          4   0   0          0    1848   2019-01-01 0:52      경기도   이천시   
4          5   0   0          0      45   2019-01-01 1:06    부산광역시  부산진구   
...      ...  ..  ..        ...     ...               ...      ...   ...   
40098  40099   0   0          0   23707  2019-12-31 22:33      경기도   파주시   
40099  40100   0   0          0    3542  2019-12-31 22:45     경상북도   경산시   
40100  40101   0   0          0       0  2019-12-31 22:47    부산광역시    남구   
40101  40102   0   0          0     105  2019-12-31 23:05  제주특별자치도   제주시   
40102  40103   0   0          0    3879  2019-12-31 23:18    인천광역시   남동구   

        읍면동         화재유형     발화열원     발화열원소분류 발화요인대분류       발화요인소분류  \
0       오분동  기타(쓰레기 화재등)  폭발물, 폭죽          폭죽     부주의          폭죽놀이   
1       안중읍       건축,구조물       미상          미상      미상            미상   
2      하월곡동     자동차,철도차량       미상          미상      미상            미상   
3       설성면  기타(쓰레기 화재등)  화학적 발화열       화학반응열  화학적 요인          자연발화   
4       당감동       건축,구조물     작동기기    기타(작동기기)  전기적 요인   절연열화에 의한 단락   
...     ...          ...      ...         ...     ...           ...   
40098   광탄면       건축,구조물     작동기기  전기적 아크(단락)  전기적 요인   접촉불량에 의한 단락   
40099   남방동       건축,구조물       미상          미상      미상            미상   
40100   문현동       건축,구조물     작동기기   기기 전도,복사열     부주의       음식물 조리중   
40101  외도일동       건축,구조물     작동기기   기기 전도,복사열     부주의      가연물 근접방치   
40102   고잔동       건축,구조물     작동기기  전기적 아크(단락)  전기적 요인  압착,손상에 의한 단락   

        최초착화물대분류           최초착화물소분류     장소대분류  장소중분류     장소소분류  
0      종이,목재,건초등             풀, 나뭇잎        기타     야외      기타야외  
1             미상                 미상        주거   단독주택      단독주택  
2             미상                 미상  자동차,철도차량    자동차     승용자동차  
3      종이,목재,건초등             풀, 나뭇잎        기타     야외       야적장  
4          전기,전자        전기, 전자기기 기판     기타서비스  기타건축물    기타 건축물  
...          ...                ...       ...    ...       ...  
40098      전기,전자               전선피복      산업시설   공장시설  제재 및 목공업  
40099  종이,목재,건초등                 건초      산업시설  동식물시설  기타 동식물시설  
40100         식품                튀김유   판매,업무시설   일반업무      오피스텔  
40101       합성수지  플라스틱, PVC, 비닐, 장판   판매,업무시설   판매시설   할인점(마트)  
40102      전기,전자               전선피복      산업시설   공장시설  전기, 전자공업  

[40103 rows x 19 columns]

In [18]:
# df_seoul = pd.read_excel('엑셀파일')
# ###데이터 변환
# df.to_csv('엑셀파일')

# # 업무자동화
# # xlsx
# # 중복제거
# # 빈 열 >> 빈 열에 어떤 값을 채워넣는다.
# # to_excel

In [20]:
df_iris = pd.read_csv('/content/sample_data/iris.csv')
df_iris

sepal.length  sepal.width  petal.length  petal.width    variety
0             5.1          3.5           1.4          0.2     Setosa
1             4.9          3.0           1.4          0.2     Setosa
2             4.7          3.2           1.3          0.2     Setosa
3             4.6          3.1           1.5          0.2     Setosa
4             5.0          3.6           1.4          0.2     Setosa
..            ...          ...           ...          ...        ...
145           6.7          3.0           5.2          2.3  Virginica
146           6.3          2.5           5.0          1.9  Virginica
147           6.5          3.0           5.2          2.0  Virginica
148           6.2          3.4           5.4          2.3  Virginica
149           5.9          3.0           5.1          1.8  Virginica

[150 rows x 5 columns]

In [22]:
df_iris.head()

sepal.length  sepal.width  petal.length  petal.width variety
0           5.1          3.5           1.4          0.2  Setosa
1           4.9          3.0           1.4          0.2  Setosa
2           4.7          3.2           1.3          0.2  Setosa
3           4.6          3.1           1.5          0.2  Setosa
4           5.0          3.6           1.4          0.2  Setosa

In [23]:
df_iris.tail()

sepal.length  sepal.width  petal.length  petal.width    variety
145           6.7          3.0           5.2          2.3  Virginica
146           6.3          2.5           5.0          1.9  Virginica
147           6.5          3.0           5.2          2.0  Virginica
148           6.2          3.4           5.4          2.3  Virginica
149           5.9          3.0           5.1          1.8  Virginica

In [24]:
df_iris.info()
# 열의이름. 총 몇행(RangeIndex)이 있는지 알수 있다.
# Non-NULL이 전체 행의 갯수와 같으면 문제가 없으나 다르면 결측치가 있다는 것이다.
# RangeIndex = Non-NULL 수가 같아서 결측치가 없다는 의미

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   sepal.length  150 non-null    float64
 1   sepal.width   150 non-null    float64
 2   petal.length  150 non-null    float64
 3   petal.width   150 non-null    float64
 4   variety       150 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.0+ KB


In [25]:
# 기술(설명, 데이터의 전반적인 모습을 설명) 통계
df_iris.describe()

sepal.length  sepal.width  petal.length  petal.width
count    150.000000   150.000000    150.000000   150.000000
mean       5.843333     3.057333      3.758000     1.199333
std        0.828066     0.435866      1.765298     0.762238
min        4.300000     2.000000      1.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.800000     3.000000      4.350000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max        7.900000     4.400000      6.900000     2.500000

In [27]:
df_titanic = pd.read_csv('/content/sample_data/titanic.csv')
df_titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [28]:
df_titanic.info()
#Age, Cabin,Embarked 결측치 존재

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [29]:
#결측치가 아닌지 확인
df_titanic.isna()

PassengerId  Survived  Pclass   Name    Sex    Age  SibSp  Parch  Ticket  \
0          False     False   False  False  False  False  False  False   False   
1          False     False   False  False  False  False  False  False   False   
2          False     False   False  False  False  False  False  False   False   
3          False     False   False  False  False  False  False  False   False   
4          False     False   False  False  False  False  False  False   False   
..           ...       ...     ...    ...    ...    ...    ...    ...     ...   
886        False     False   False  False  False  False  False  False   False   
887        False     False   False  False  False  False  False  False   False   
888        False     False   False  False  False   True  False  False   False   
889        False     False   False  False  False  False  False  False   False   
890        False     False   False  False  False  False  False  False   False   

      Fare  Cabin  Embarked  
0    False   True     False  
1    False  False     False  
2    False   True     False  
3    False  False     False  
4    False   True     False  
..     ...    ...       ...  
886  False   True     False  
887  False  False     False  
888  False   True     False  
889  False  False     False  
890  False   True     False  

[891 rows x 12 columns]

In [30]:
#결측치가 몇개가 있는가
df_titanic.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [31]:
# 데이터 프레임에서 한 열만 조회
# df_titanic['열이름']

# 새 열 추가
# 기존 열이 없다면 새 열이 만들어짐
# df_titanic['새 열이름']
df_titanic['만나이'] = df_titanic['Age']-1
df_titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked   만나이  
0      0         A/5 21171   7.2500   NaN        S  21.0  
1      0          PC 17599  71.2833   C85        C  37.0  
2      0  STON/O2. 3101282   7.9250   NaN        S  25.0  
3      0            113803  53.1000  C123        S  34.0  
4      0            373450   8.0500   NaN        S  34.0

결측치 해결법
1. (제일 쉬운 방법)열 삭제 df.drop: 결측치가 너무 많은 열
   (데이터 분석을 현실적으로 할수 있겠느냐)
    ex. Cabin(선실 등급) 900명 중 687건 결측치
2. 행 삭제 dropna: 결측치가 적어서 데이터 분석에 영향을 그닥 미치지 않을때
   ex. Embarked(중간 정착치) : 전체 900명 중 2건만 결측치
3.임의의 값으로 채워넣기 fillna
  > 평균(대표값)
  > 고정된 문자(ex. 알수없음)
  > 이전 값
4. 예측해서 채워넣기

In [32]:
df_titanic['Age'] = df_titanic['Age'].fillna(df_titanic['Age'].mean())
df_titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked   만나이  
0      0         A/5 21171   7.2500   NaN        S  21.0  
1      0          PC 17599  71.2833   C85        C  37.0  
2      0  STON/O2. 3101282   7.9250   NaN        S  25.0  
3      0            113803  53.1000  C123        S  34.0  
4      0            373450   8.0500   NaN        S  34.0

In [33]:
df_titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 13 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          891 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
 12  만나이          714 non-null    float64
dtypes: float64(3), int64(5), object(5)
memory usage: 90.6+ KB


1.LLM 구조화된 프롬프트 작성


*   아래 작업을 수행하는 python 코드를 작성해줘
*   데이터 정보 제공
    - Age: 나이
    - Fare: 요금
*   동작설명
    - 1.
    - 2.    
    - 3.
    - 4.

2.지시어로만 되어 있는 프롬프트 : 부적합하다.
*   ~에 대해 알려줘
*   000에 대해서 글을 작성해줘

df_titanic 데이터프레임을 이용해 다음 작업을 수행하는 python코드를 작성해줘.

# 동작설명
1. 다음 열들을 삭제
*   PassengerId
*   Name
*  Ticket
2. SibSp열과	Parch열의 값을 더해 Family라는 새로운 열 생성
3.Age열의 값이 결측치인 경우 Age열의 평균으로 대체
4.Cabin열의 값이 결측치인 경우 X라는 문자로 대체
5.Embarked 열의 값이 결측치인 경우 해당 행을 삭제
6.Cabin열 열에서 첫 글자만 Cabin2라는 열로 새로 입력


In [34]:
df_titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked   만나이  
0      0         A/5 21171   7.2500   NaN        S  21.0  
1      0          PC 17599  71.2833   C85        C  37.0  
2      0  STON/O2. 3101282   7.9250   NaN        S  25.0  
3      0            113803  53.1000  C123        S  34.0  
4      0            373450   8.0500   NaN        S  34.0

In [40]:


df_titanic = pd.read_csv('/content/sample_data/titanic.csv') # Add this line to re-initialize df_titanic

# 1) 열 삭제
df_titanic = df_titanic.drop(columns=["PassengerId", "Name", "Ticket"])

# 2) SibSp + Parch => Family 열 생성
df_titanic["Family"] = df_titanic["SibSp"] + df_titanic["Parch"]

# 3) Age 결측치 => Age 평균으로 대체
age_mean = df_titanic["Age"].mean()
df_titanic["Age"] = df_titanic["Age"].fillna(age_mean)

# 4) Cabin 결측치 => 'X'로 대체
df_titanic["Cabin"] = df_titanic["Cabin"].fillna("X")

# 5) Embarked 결측치 행 삭제
df_titanic = df_titanic.dropna(subset=["Embarked"])

# 6) Cabin 첫 글자만 추출하여 Cabin2 열 생성
#    (Cabin이 'X'인 경우 Cabin2도 'X'가 됨)
df_titanic["Cabin2"] = df_titanic["Cabin"].astype(str).str[0]
df_titanic.head()

Survived  Pclass     Sex   Age  SibSp  Parch     Fare Cabin Embarked  \
0         0       3    male  22.0      1      0   7.2500     X        S   
1         1       1  female  38.0      1      0  71.2833   C85        C   
2         1       3  female  26.0      0      0   7.9250     X        S   
3         1       1  female  35.0      1      0  53.1000  C123        S   
4         0       3    male  35.0      0      0   8.0500     X        S   

   Family Cabin2  
0       1      X  
1       1      C  
2       0      X  
3       1      C  
4       0      X